### ISSUE TITLE
Fully implement flexbox

### ISSUE THREAD

**Issue by shinglyu** [servo/servo#12453]
This is a meta bug to track flexbox implementation progress. 

Here is a popularity metrics by Chrome and Edge we can reference:

| Property | Rank | Chrome % | Bing % | Partially Implemented | Remaing Failed Tests |
| --- | --- | --- | --- | --- | --- |
| flex | 19 | 18.83% |  | Yes | `inline-flex` |
| alias-webkit-flex | 12 | 12.32% |  |  |  |
| flex-direction | 12 | 11.82% | 1.70% | Yes | 3 |
| flex-grow | 11 | 10.66% | 1.50% |  |  |
| webkit-box-flex | 10 | 9.56% | 0.57% |  |  |
| flex-wrap | 7 | 6.59% | 1.60% |  |  |
| alias-webkit-flex-direction | 6 | 6.33% |  |  |  |
| alias-webkit-flex-grow | 5 | 4.52% |  |  |  |
| alias-webkit-flex-wrap | 5 | 5.17% |  |  |  |
| alias-webkit-flex-flow | 4 | 4.10% |  |  |  |
| flex-shrink | 4 | 4.04% | 1.40% |  |  |
| alias-webkit-flex-shrink | 3 | 2.69% |  |  |  |
| flex-basis | 3 | 3.38% | 1.30% |  |  |
| flex-flow | 3 | 3.33% |  |  |  |
| alias-webkit-flex-basis | 2 | 1.90% |  |  |  |
| webkit-box-flex-group | 0 | 0.04% | 0.00% |  |  |
| align-self |  |  |  |  | `stretch` |

**Comment by stshine** [comment]
Hi @shinglyu , Glad you are interested! I have some working code that I believe have implemented a lot of features of flexbox(currently in row container only):
https://github.com/stshine/servo/tree/flexbox
Currently my PR submit is stuck at #12330 . Would you like to have a look at it before you start to work on flexbox?

**Comment by shinglyu** [comment]
@stshine Thank you! I'll take a look next Monday. If you need to find a reviewer with less review in the queue, I can probably help.

**Comment by stshine** [comment]
Definitely! I do hope things could go faster :)

**Comment by paulrouget** [comment]
Related: https://github.com/servo/servo/issues/7304

**Comment by shinglyu** [comment]
@stshine : Seems like you have big plans for flexbox, would you mind sharing your plan here?

**Comment by stshine** [comment]
You can take a look at my tree, it is an implementation of most basic flexbox features :) After I submit that I will reply here with some remaining important features I gathered and the technical difficulties to implement them.

**Comment by shinglyu** [comment]
A few websites we can use as benchmark, suggested by Soledad Penadés and Belén Albeza:

* https://philipwalton.github.io/solved-by-flexbox/ 
  * It is a very popular website that people refer a lot to. The CSS code for each showcase is at the bottom of the page too so you can see which properties are in use.
* https://css-tricks.com/snippets/css/a-guide-to-flexbox/
 * "the guide at CSS tricks is what everyone uses to build their flexbox layouts"

**Comment by shinglyu** [comment]
@stshine Your work is pretty impressive! I'm still reading the whole patch, is there any feature/fix I can do in parallel when you are preparing the patch?

**Comment by stshine** [comment]
Glad your appreciate it :laughing: 
One thing you may interested is [Implied Minimum Size of Flex Items](https://drafts.csswg.org/css-flexbox/#min-size-auto). To do this you may need to change the type of `min-{width, height}` in `position.mako.rs` to `LengthOrPercentageOrAuto`. Then in `flex.rs' when the computed value of `min_inline_size` is `Auto` and `overflow` is `visible`, set the `min_size` to a implied size according to the rules. No need to deal with minimal block size yet.
This is not tightly relevant to flexbox, but to make it functional it still needs #12230 merged first. I will have my patch prepared in one or two days.

**Comment by stshine** [comment]
#12330 is ready :) Since I am not experienced, Please give it some criticisms, especially  on the last commit to help me eliminate all the stupidty out!

**Comment by shinglyu** [comment]
Thanks! I'll look into the implied minimum size stuff then.

**Comment by stshine** [comment]
Since #12330 is merged, here is some significant missing features I've collected. Still a lot is left.

- multi-line column flexbox and `align-self: stretch`. 

This is the most troublesome one.  Implementing this require functionalities of: 1. explicitly reflow a flex item before other flows, 2. make a flex item of auto size behave like it has definite size. I will discuss this in a new issue.

- painting order of flex item. 

As the spec in https://drafts.csswg.org/css-flexbox/#order-modified-document-order , the painting order should also follow the `order` property, starting from the lowest to highest. However, with my glance at https://github.com/servo/servo/blob/master/components/layout/traversal.rs#L211 it looks like the order of display list is solely determined by the document order. One way to solve this is change the traverse() of `BuildDisplayList` so that when it meets a flexflow it will use the `items` field of to change the display list order, but this would require a vtable look up for each flow, which may slow down servo a lot. Another way is to sort the flow children of a flexflow directly, but 1. rust does not provide sort for linked list, and currently I don't know how to write a efficienct one for flowlist, 2. the spec requires the `order` property to change only the display order and painting order of elements, and nothing else, which may not be valid if sorting to flow tree. A better solution may be sorting the display lists before they are sending to renderer, but I need to do more investigation for better performance.

- layout of absolutely-positioned flex children

The spec for this is at https://drafts.csswg.org/css-flexbox/#abspos-items :
> The static position of an absolutely-positioned child of a flex container is determined such that the child is positioned as if it were the sole flex item in the flex container, assuming both the child and the flex container were fixed-size boxes of their used size.

But I have to say I can't get the concise meaning, while Blink and Gecko behave differently on this. Diving into tests maybe helpful.

- flex item with intrinsic aspect ratio

Item like image that have its intrinsic ratio needs special handling to keep its ratio after flex lengths resolving. Currently servo handle them in https://github.com/servo/servo/blob/master/components/layout/fragment.rs#L488 , which get the size info directly from style, thus won't be fit into flexbox. That method needs to be changed, maybe isolating logic for intrinsic aspect ratio is good idea. 

- Proper implementation of `bubble_inline_size` of flex container

The spec link is https://drafts.csswg.org/css-flexbox/#intrinsic-main-sizes , but both Blink and Gecko do not behave property, especially on percentage margins and paddings , So I assume this should not be something of high priority. A naive implementation but better than the current one maybe helpful.

- display: inline-flex

An inline flex flow should behave like inline-block in normal layout, as the flex flow like block. This should not be difficult after studying the implementation of inline block, but it relies more on the implementation of `bubble_inline_size`.

- baseline alignment

This requires servo to implement baseline for all block-like flow first.

- support for `visibility: collapse`

This require servo to support `visibility: collapse` first, which has not been done in my glance.

- Implied minimum size 

@shinglyu has interest on this, see #12610 for detail.

**Comment by shinglyu** [comment]
@stshine Thank you for sorting that out! It's a really remarkable work! 

I'm currently working  on https://github.com/servo/servo/issues/12676, which blocks some of the flexbox reftests, after that lands I'll get back and work on the flexbox.

**Comment by shinglyu** [comment]
During the [Web Engines Hackfest], someone suggested us to run the [Chromium flexbox test](https://cs.chromium.org/chromium/src/third_party/WebKit/LayoutTests/css3/flexbox/) alongside the incomplete wpt flexbox test.

**Comment by shinglyu** [comment]
`display:inline-flex` implemented in ##14978

**Comment by shinglyu** [comment]
I'm also looking at writing-mode reftests, but looks like there are some releated issue with #14144 and #10479

**Comment by nox** [comment]
How fare are we on this?

**Comment by jdm** [comment]
Lots more flexbox work required.

**Comment by gnzlbg** [comment]
It might be worth it to use [stretch](https://github.com/vislyhq/stretch) here.

**Comment by mrobinson** [comment]
I think the majority of flexbox is implemented at this point so the metabug is probably not so useful. The main missing features are orthogonal writing mode support (#2560) and `visibility: collapse` (#20142).
